addition of ast.Constant type makes astroid crash on import when using current CPython master (which will become 3.8)

It looks like the merge of https://github.com/python/cpython/pull/9445 broke astroid on CPython 3.8 pre-releases:

```
../../../../virtualenv/python3.8-dev/lib/python3.8/site-packages/trio/tests/test_exports.py:9: in <module>
    from pylint.lint import PyLinter
../../../../virtualenv/python3.8-dev/lib/python3.8/site-packages/pylint/lint.py:75: in <module>
    import astroid
../../../../virtualenv/python3.8-dev/lib/python3.8/site-packages/astroid/__init__.py:156: in <module>
    __import__(module[:-3])
../../../../virtualenv/python3.8-dev/lib/python3.8/site-packages/astroid/brain/brain_builtin_inference.py:103: in <module>
    extend_builtins({'bytes': partial(_extend_str, rvalue="b''"),
../../../../virtualenv/python3.8-dev/lib/python3.8/site-packages/astroid/brain/brain_builtin_inference.py:99: in extend_builtins
    transform(builtin_ast[class_name])
../../../../virtualenv/python3.8-dev/lib/python3.8/site-packages/astroid/brain/brain_builtin_inference.py:84: in _extend_str
    fake = AstroidBuilder(MANAGER).string_build(code)['whatever']
../../../../virtualenv/python3.8-dev/lib/python3.8/site-packages/astroid/builder.py:132: in string_build
    module = self._data_build(data, modname, path)
../../../../virtualenv/python3.8-dev/lib/python3.8/site-packages/astroid/builder.py:173: in _data_build
    module = builder.visit_module(node, modname, node_file, package)
../../../../virtualenv/python3.8-dev/lib/python3.8/site-packages/astroid/rebuilder.py:162: in visit_module
    newnode.postinit([self.visit(child, newnode) for child in node.body])
../../../../virtualenv/python3.8-dev/lib/python3.8/site-packages/astroid/rebuilder.py:162: in <listcomp>
    newnode.postinit([self.visit(child, newnode) for child in node.body])
../../../../virtualenv/python3.8-dev/lib/python3.8/site-packages/astroid/rebuilder.py:174: in visit
    return visit_method(node, parent)
../../../../virtualenv/python3.8-dev/lib/python3.8/site-packages/astroid/rebuilder.py:954: in visit_classdef
    return super(TreeRebuilder3, self).visit_classdef(node, parent,
../../../../virtualenv/python3.8-dev/lib/python3.8/site-packages/astroid/rebuilder.py:394: in visit_classdef
    [self.visit(child, newnode)
../../../../virtualenv/python3.8-dev/lib/python3.8/site-packages/astroid/rebuilder.py:394: in <listcomp>
    [self.visit(child, newnode)
../../../../virtualenv/python3.8-dev/lib/python3.8/site-packages/astroid/rebuilder.py:174: in visit
    return visit_method(node, parent)
../../../../virtualenv/python3.8-dev/lib/python3.8/site-packages/astroid/rebuilder.py:576: in visit_functiondef
    return self._visit_functiondef(nodes.FunctionDef, node, parent)
../../../../virtualenv/python3.8-dev/lib/python3.8/site-packages/astroid/rebuilder.py:566: in _visit_functiondef
    body=[self.visit(child, newnode) for child in node.body],
../../../../virtualenv/python3.8-dev/lib/python3.8/site-packages/astroid/rebuilder.py:566: in <listcomp>
    body=[self.visit(child, newnode) for child in node.body],
../../../../virtualenv/python3.8-dev/lib/python3.8/site-packages/astroid/rebuilder.py:174: in visit
    return visit_method(node, parent)
../../../../virtualenv/python3.8-dev/lib/python3.8/site-packages/astroid/rebuilder.py:741: in visit_return
    newnode.postinit(self.visit(node.value, newnode))
../../../../virtualenv/python3.8-dev/lib/python3.8/site-packages/astroid/rebuilder.py:172: in visit
    visit_method = getattr(self, visit_name)
E   AttributeError: 'TreeRebuilder3' object has no attribute 'visit_constant'
```

(See: https://travis-ci.org/python-trio/trio/jobs/435176057)

--- a/ChangeLog
+++ b/ChangeLog
@@ -34,6 +34,8 @@ Release Date: TBA
 
      Close PyCQA/pylint#2434
 
+   * Make compatible with AST changes in Python 3.8.
+
 
 What's New in astroid 2.0.4?
 ============================

--- a/astroid/as_string.py
+++ b/astroid/as_string.py
@@ -188,6 +188,8 @@ class AsStringVisitor:
 
     def visit_const(self, node):
         """return an astroid.Const node as string"""
+        if node.value is Ellipsis:
+            return '...'
         return repr(node.value)
 
     def visit_continue(self, node):
@@ -273,7 +275,7 @@ class AsStringVisitor:
 
     def visit_extslice(self, node):
         """return an astroid.ExtSlice node as string"""
-        return ','.join([dim.accept(self) for dim in node.dims])
+        return ', '.join([dim.accept(self) for dim in node.dims])
 
     def visit_for(self, node):
         """return an astroid.For node as string"""
@@ -430,8 +432,15 @@ class AsStringVisitor:
 
     def visit_subscript(self, node):
         """return an astroid.Subscript node as string"""
-        return '%s[%s]' % (self._precedence_parens(node, node.value),
-                           node.slice.accept(self))
+        idx = node.slice
+        if idx.__class__.__name__.lower() == 'index':
+            idx = idx.value
+        idxstr = idx.accept(self)
+        if idx.__class__.__name__.lower() == 'tuple' and idx.elts:
+            # Remove parenthesis in tuple and extended slice.
+            # a[(::1, 1:)] is not valid syntax.
+            idxstr = idxstr[1:-1]
+        return '%s[%s]' % (self._precedence_parens(node, node.value), idxstr)
 
     def visit_tryexcept(self, node):
         """return an astroid.TryExcept node as string"""

--- a/astroid/rebuilder.py
+++ b/astroid/rebuilder.py
@@ -479,6 +479,7 @@ class TreeRebuilder:
         newnode.postinit(self.visit(node.value, newnode))
         return newnode
 
+    # Not used in Python 3.8+.
     def visit_ellipsis(self, node, parent):
         """visit an Ellipsis node by returning a fresh instance of it"""
         return nodes.Ellipsis(getattr(node, 'lineno', None),
@@ -509,6 +510,7 @@ class TreeRebuilder:
                          _visit_or_none(node, 'locals', self, newnode))
         return newnode
 
+    # Not used in Python 3.8+.
     def visit_extslice(self, node, parent):
         """visit an ExtSlice node by returning a fresh instance of it"""
         newnode = nodes.ExtSlice(parent=parent)
@@ -641,6 +643,7 @@ class TreeRebuilder:
             parent.set_local(name.split('.')[0], newnode)
         return newnode
 
+    # Not used in Python 3.8+.
     def visit_index(self, node, parent):
         """visit a Index node by returning a fresh instance of it"""
         newnode = nodes.Index(parent=parent)
@@ -702,12 +705,19 @@ class TreeRebuilder:
             self._save_assignment(newnode)
         return newnode
 
+    def visit_constant(self, node, parent):
+        """visit a Constant node by returning a fresh instance of Const"""
+        return nodes.Const(node.value, getattr(node, 'lineno', None),
+                           getattr(node, 'col_offset', None), parent)
+
+    # Not used in Python 3.8+.
     def visit_str(self, node, parent):
         """visit a String/Bytes node by returning a fresh instance of Const"""
         return nodes.Const(node.s, getattr(node, 'lineno', None),
                            getattr(node, 'col_offset', None), parent)
     visit_bytes = visit_str
 
+    # Not used in Python 3.8+.
     def visit_num(self, node, parent):
         """visit a Num node by returning a fresh instance of Const"""
         return nodes.Const(node.n, getattr(node, 'lineno', None),
@@ -854,6 +864,7 @@ class TreeRebuilder3(TreeRebuilder):
         """visit an arg node by returning a fresh AssName instance"""
         return self.visit_assignname(node, parent, node.arg)
 
+    # Not used in Python 3.8+.
     def visit_nameconstant(self, node, parent):
         # in Python 3.4 we have NameConstant for True / False / None
         return nodes.Const(node.value, getattr(node, 'lineno', None),

--- a/astroid/tests/unittest_nodes.py
+++ b/astroid/tests/unittest_nodes.py
@@ -158,8 +158,8 @@ def function(var):
         self.assertEqual(ast.as_string(), 'a[...]')
 
     def test_slices(self):
-        for code in ('a[0]', 'a[1:3]', 'a[:-1:step]', 'a[:,newaxis]',
-                     'a[newaxis,:]', 'del L[::2]', 'del A[1]', 'del Br[:]'):
+        for code in ('a[0]', 'a[1:3]', 'a[:-1:step]', 'a[:, newaxis]',
+                     'a[newaxis, :]', 'del L[::2]', 'del A[1]', 'del Br[:]'):
             ast = abuilder.string_build(code).body[0]
             self.assertEqual(ast.as_string(), code)
 
@@ -170,7 +170,7 @@ del bree[3:d]
 bord[2:]
 del av[d::f], a[df:]
 a[:1] = bord[2:]
-del SRC[::1,newaxis,1:]
+del SRC[::1, newaxis, 1:]
 tous[vals] = 1010
 del thousand[key]
 del a[::2], a[:-1:step]
@@ -546,7 +546,7 @@ class AnnAssignNodeTest(unittest.TestCase):
             print()
             test: int = 5
             test2: str
-            test3: List[Dict[(str, str)]] = []
+            test3: List[Dict[str, str]] = []
         """)
         ast = abuilder.string_build(code)
         self.assertEqual(ast.as_string().strip(), code.strip())

